### PR TITLE
Cmd authorization support

### PR DESCRIPTION
**Pull request by jmthomas**
This is a lot to review but I'd appreciate a close look at the new comment blocks because I was describing things that I found out along the way.

Also there is probably additional testing needed and some frontend changes coming.

### CODE REVIEW

File: openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue [hunk 1]
```diff
@@ -63,25 +149,149 @@ export default {
...
+      Api.post('/openc3-api/cmdauth/release', {
+        data: {
+          target_name: target_name,
+        },
+      })
+    },
+    takeAll() {
+      Api.post('/openc3-api/cmdauth/take-all', {
+        data: {
+          target_names: this.data.map((target) => target.name),
```
Comment (ryanmelt) [inline openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue:284]: Why does this need a list of target names?

File: openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue [hunk 2]
```diff
@@ -63,25 +149,149 @@ export default {
...
+        },
+      })
+    },
+    releaseAll() {
+      Api.post('/openc3-api/cmdauth/release-all', {
+        data: {
+          target_names: this.data.map((target) => target.name),
```
Comment (ryanmelt) [inline openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue:291]: Ditto.

File: openc3/lib/openc3/api/cmd_api.rb [hunk 3]
```diff
@@ -173,7 +177,7 @@ def send_raw(interface_name, data, scope: $openc3_scope, token: $openc3_token)
     # @param target_name [String] Target name of the command
     # @param command_name [String] Packet name of the command
     # @return [Hash] command hash with last command buffer
-    def get_cmd_buffer(*args, scope: $openc3_scope, token: $openc3_token)
+    def get_cmd_buffer(*args, manual: false, scope: $openc3_scope, token: $openc3_token)
       target_name, command_name = _extract_target_command_names('get_cmd_buffer', *args)
       authorize(permission: 'cmd_info', target_name: target_name, packet_name: command_name, scope: scope, token: token)
```
Comment (ryanmelt) [inline openc3/lib/openc3/api/cmd_api.rb:182]: Any harm in passing manual down to authorize always?

File: openc3/python/tester.py [hunk 4]
```diff
@@ -0,0 +1 @@
+print('Hello World')
```
Comment (ryanmelt) [inline openc3/python/tester.py:1]: This file is probably accidentally here?

File: openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js [hunk 5]
```diff
@@ -47,7 +47,7 @@ class Auth {
     return { name: 'Anonymous' }
   }
   userroles() {
-    return ['ALLSCOPES__admin']
+    return ['admin']
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js:50]: Did you change the implementation of this in Enterprise to do all the filtering and such?

File: openc3/lib/openc3/topics/system_topic.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# This file may also be used under the terms of a commercial license
+# if purchased from OpenC3, Inc.
+
+require 'openc3/topics/topic'
+
+module OpenC3
+  class SystemTopic < Topic
```
Comment (ryanmelt) [inline openc3/lib/openc3/topics/system_topic.rb:22]: Why not SystemEventsTopic?

**Review by ryanmelt — CHANGES_REQUESTED** [review]
All pretty minor comments.

File: openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmthomas) [reply openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js:50]: Yes, the new implementation in Enterprise just returns the roles without the scope attached so this change matches the new implementation. I found that we were repeating this logic in 3 different places.

File: openc3/python/tester.py (hunk 4 above)
Comment (jmthomas) [reply openc3/python/tester.py:1]: Will delete

File: openc3/lib/openc3/topics/system_topic.rb (hunk 6 above)
Comment (jmthomas) [reply openc3/lib/openc3/topics/system_topic.rb:22]: Will change

File: openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue (hunk 1 above)
Comment (jmthomas) [reply openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue:284]: Probably doesn't but I already had the list in the frontend. Probably more robust to ask for the list in the backend.

File: openc3-cosmos-cmd-tlm-api/app/controllers/application_controller.rb [hunk 7]
```diff
@@ -39,10 +39,11 @@ def username()
...
   # Authorize and rescue the possible execeptions
   # @return [Boolean] true if authorize successful
-  def authorization(permission)
+  def authorization(permission, target_name: nil)
     begin
       authorize(
         permission: permission,
+        target_name: target_name,
         scope: params[:scope],
         token: request.headers['HTTP_AUTHORIZATION'],
```
[the remark this answers is not in the document]
Comment (jmthomas) [reply openc3-cosmos-cmd-tlm-api/app/controllers/application_controller.rb:49]: Will add

File: openc3-cosmos-cmd-tlm-api/config/initializers/cts.rb [hunk 8]
```diff
@@ -27,6 +27,10 @@
 
 module OpenC3
   class Cts
+    # This sets a global flag $openc3_authorize = true
+    # which is used by authorization.rb to bypass the
+    # role and permission checks. This is because the Cts
+    # is an internal microservice inside the trust zone.
```
[the remark this answers is not in the document]
Comment (jmthomas) [reply openc3-cosmos-cmd-tlm-api/config/initializers/cts.rb:33]: Fixed

File: openc3/lib/openc3/api/cmd_api.rb (hunk 3 above)
Comment (jmthomas) [reply openc3/lib/openc3/api/cmd_api.rb]: I moved the logic all into the Enterprise authorize method like we talked about

**Review by ryanmelt — APPROVED** [review]

**Pull request merged**